dbms/src/IO/SharedReadBuffer.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <Formats/FormatFactory.h>
+#include <IO/BufferWithOwnMemory.h>
+#include <IO/ReadBuffer.h>
+
+
+namespace DB
+{
+
+/** Allows many threads to read from single ReadBuffer
+  * Each SharedReadBuffer has his own memory,
+  *  which he filles under shared mutex using FileSegmentationEngine.
+  */
+class SharedReadBuffer : public BufferWithOwnMemory<ReadBuffer>
+{
+public:
+    SharedReadBuffer(
+            ReadBuffer & in_,
+            std::shared_ptr<std::mutex> & mutex_,
+            FormatFactory::FileSegmentationEngine file_segmentation_engine_,
+            size_t min_chunk_size_)
+        : BufferWithOwnMemory<ReadBuffer>(),
+        mutex(mutex_),
+        in(in_),
+        file_segmentation_engine(file_segmentation_engine_),
+        min_chunk_size(min_chunk_size_)
+    {
+    }
+
+private:
+    bool nextImpl() override
+    {
+        if (eof || !mutex)
+            return false;
```

Comment:
maybe in the case of null mutex  also write some `LOG_ERROR`?
What do you think @KochetovNicolai ?